src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -85,13 +102,19 @@ public void TryKickOffPrimaryBranchWork_NoLock(Func<Solution, CancellationToken,
                 if (_primaryBranchTask != null)
                     return;
 
-                _primaryBranchTask = ComputePrimaryBranchAsync();
+                // Grab the cancellation token up front while we know we are holding the lock and mutating state.  We
+                // don't want to potentially grab it at some undefined point at the future when this type has already
+                // had its in-flight-count reduced to 0 and thus had our CTS be disposed.
+                var token = this.CancellationToken;
+                _primaryBranchTask = ComputePrimaryBranchAsync(token);
                 return;
 
-                async Task<Solution> ComputePrimaryBranchAsync()
+                async Task<Solution> ComputePrimaryBranchAsync(CancellationToken cancellationToken)
                 {
                     var solution = await _disconnectedSolutionTask.ConfigureAwait(false);
-                    return await updatePrimaryBranchAsync(solution, _cancellationTokenSource.Token).ConfigureAwait(false);
```

Comment:
this would could absolutely start at some arbitrary point in the future, and definitely after a point that hte cts had been cancelled/disposed.  accessing .Token off of a Disposed cts will throw an exception.  If _disconnectedSolutionTask was complete, then that could happen synchronously here inside TryKickOffPrimaryBranchWork_NoLock which is called in a method which must run to completion to setup state properly.

THat said... i also have a feeling the synchronous case cannot be hit, because we shouldn't be able to observe a cancelled token source synchronously since we're in the same lock that is used to cancel the cts in the first place.